### PR TITLE
Support syncing HA Areas to HomeKit Rooms

### PR DESCRIPTION
**Pull request by robbiet480**
Depends on https://github.com/home-assistant/core/pull/37218

### CODE REVIEW

**Comment by robbiet480** [comment]
@zacwest Mind taking a look at this before I continue to flesh it out? The data model feels a bit wrong but not really sure how. Obviously I need to get this hooked up to the webhook too, will do that once you sign off on the idea and implementation.

Thanks!

**Comment by Murua83Skye** [comment]
``_ San Jose, California_``**my weview/desktop iOS,iPhoneSE.moble,device.network**[Uploading 85982011...]()**_**

**Comment by robbiet480** [comment]
@Murua83Skye ?

File: HomeAssistant/Views/SettingsDetailViewController.swift [hunk 1]
```diff
@@ -781,6 +845,172 @@ class SettingsDetailViewController: FormViewController, TypedRowControllerType {
             }
         }
     }
+
+    func getSerialNumberForAccessory(_ accessory: HMAccessory) -> String? {
+        let infoService = accessory.services.first { $0.serviceType == HMServiceTypeAccessoryInformation }
+
+        let serialNumberChar = infoService?.characteristics.first {
+            $0.characteristicType == "00000030-0000-1000-8000-0026BB765291"
```
Comment (zacwest) [inline HomeAssistant/Views/SettingsDetailViewController.swift:853]: `HMCharacteristicTypeSerialNumber` looks deprecated, is there another way to get this info?

File: HomeAssistant/Views/SettingsDetailViewController.swift [hunk 2]
```diff
@@ -781,6 +845,172 @@ class SettingsDetailViewController: FormViewController, TypedRowControllerType {
...
+        })
+
+        print("roomMap", roomMap)
+
+        print("accessoryMap", accessoryMap)
+
+        guard let areaName = testObj["name"] as? String,
+            let entitiesArr = testObj["entities"] as? [String] else { return nil }
+
+        for entityId in entitiesArr {
```
Comment (zacwest) [inline HomeAssistant/Views/SettingsDetailViewController.swift:945]: This may also be a great opportunity to use `Dictionary(grouping:by:)`

Comment (zacwest) [reply HomeAssistant/Views/SettingsDetailViewController.swift:945]:
```
struct Assignment {
  let accessory: HMAccessory
  let isNew: Bool
}

enum Room: Hashable {
  case existing(HMRoom)
  case new(String)
}

Dictionary<Room, Assignment>
```

File: HomeAssistant/Views/SettingsDetailViewController.swift (hunk 1 above)
Comment (robbiet480) [reply HomeAssistant/Views/SettingsDetailViewController.swift:853]: I checked and couldn’t find anything better sadly.

**Comment by robbiet480** [comment]
Depends on #772

**Comment by jyz0501** [comment]
this work is very prefect！

**Review by Jesterday02 — APPROVED** [review]

**Review by Jesterday02 — APPROVED** [review]

**Review by Jesterday02 — APPROVED** [review]

**Pull request closed without merging**